File: C#/WindowsPhone/LGame-XNA-lib/Loon.Action.Sprite/StatusBars.cs
using Loon.Core;
using System.Collections.Generic;
using Loon.Utils;
using Loon.Core.Graphics.Opengl;
using Loon.Core.Geom;
namespace Loon.Action.Sprite {
	
	public sealed class StatusBars : LObject, ISprite {
	
		private List<StatusBar> barCaches;
	
		private bool visible;
	
		public StatusBars() {
            this.barCaches = new List<StatusBar>(
					CollectionUtils.INITIAL_CAPACITY);
			this.visible = true;
		}
	
		public StatusBar AddBar(int value_ren, int maxValue, int x, int y, int w, int h) {
			 lock (barCaches) {
						StatusBar bar = new StatusBar(value_ren, maxValue, x, y, w, h);
						CollectionUtils.Add(barCaches,bar);
						return bar;
					}
		}
	
		public StatusBar AddBar(int x, int y, int width, int height) {
			return AddBar(100, 100, x, y, width, height);
		}
	
		public StatusBar AddBar(int width, int height) {
			return AddBar(100, 100, 0, 0, width, height);
		}
	
		public void AddBar(StatusBar bar) {
			 lock (barCaches) {
						CollectionUtils.Add(barCaches,bar);
					}
		}
	
		public bool RemoveBar(StatusBar bar) {
			if (bar == null) {
				return false;
			}
			 lock (barCaches) {
						return CollectionUtils.Remove(barCaches,bar);
					}
		}
	
		public int Size() {
			return barCaches.Count;
		}
	
		public void Clear() {
			 lock (barCaches) {
						CollectionUtils.Clear(barCaches);
					}
		}
	
		public void Hide(StatusBar bar) {
			if (bar != null) {
				bar.SetVisible(false);
			}
		}
	
		public void Show(StatusBar bar) {
			if (bar != null) {
				bar.SetVisible(true);
			}
		}
	
		public void CreateUI(GLEx g) {
			if (!visible) {
				return;
			}
			int size = barCaches.Count;
			if (size > 0) {
				 lock (barCaches) {
								StatusBar.GLBegin();
								for (int i = 0; i < size; i++) {
									StatusBar bar = barCaches[i];
									if (bar != null && bar.visible) {
										bar.CreateUI(g);
									}
								}
								StatusBar.GLEnd();
							}
			}
		}

        public override void Update(long elapsedTime)
        {
			if (!visible) {
				return;
			}
			int size = barCaches.Count;
			if (size > 0) {
				 lock (barCaches) {
								for (int i = 0; i < size; i++) {
									StatusBar bar = barCaches[i];
									if (bar != null && bar.visible) {
										bar.Update(elapsedTime);
									}
								}
							}
			}
		}
	
		public void SetVisible(bool v) {
			this.visible = v;
		}
	
		public bool IsVisible() {
			return visible;
		}
	
		public RectBox GetCollisionBox() {
			return null;
		}
	
		public LTexture GetBitmap() {
			return null;
		}

        public override int GetWidth()
        {
			return 0;
		}

        public override int GetHeight()
        {
			return 0;
		}
	
		public void Dispose() {
			this.visible = false;
			int size = barCaches.Count;
			for (int i = 0; i < size; i++) {
				StatusBar bar = barCaches[i];
				if (bar != null) {
					bar.Dispose();
					bar = null;
				}
			}
			CollectionUtils.Clear(barCaches);
			barCaches = null;
		}
	}
}

File: Java/OpenGL-1.0(old_ver)/Loon-backend-Android/src/loon/action/sprite/StatusBars.java
/**
 * Copyright 2008 - 2012
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.3.3
 */
package loon.action.sprite;

import java.util.ArrayList;

import loon.core.LObject;
import loon.core.geom.RectBox;
import loon.core.graphics.opengl.GLEx;
import loon.core.graphics.opengl.LTexture;
import loon.utils.CollectionUtils;


public final class StatusBars extends LObject implements ISprite {

	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;

	private ArrayList<StatusBar> barCaches;

	private boolean visible;

	public StatusBars() {
		this.barCaches = new ArrayList<StatusBar>(
				CollectionUtils.INITIAL_CAPACITY);
		this.visible = true;
	}

	public StatusBar addBar(int value, int maxValue, int x, int y, int w, int h) {
		synchronized (barCaches) {
			StatusBar bar = new StatusBar(value, maxValue, x, y, w, h);
			barCaches.add(bar);
			return bar;
		}
	}

	public StatusBar addBar(int x, int y, int width, int height) {
		return addBar(100, 100, x, y, width, height);
	}

	public StatusBar addBar(int width, int height) {
		return addBar(100, 100, 0, 0, width, height);
	}

	public void addBar(StatusBar bar) {
		synchronized (barCaches) {
			barCaches.add(bar);
		}
	}

	public boolean removeBar(StatusBar bar) {
		if (bar == null) {
			return false;
		}
		synchronized (barCaches) {
			return barCaches.remove(bar);
		}
	}
	
	public int size(){
		return barCaches.size();
	}

	public void clear() {
		synchronized (barCaches) {
			barCaches.clear();
		}
	}

	public void hide(StatusBar bar) {
		if (bar != null) {
			bar.setVisible(false);
		}
	}

	public void show(StatusBar bar) {
		if (bar != null) {
			bar.setVisible(true);
		}
	}

	@Override
	public void createUI(GLEx g) {
		if (!visible) {
			return;
		}
		int size = barCaches.size();
		if (size > 0) {
			synchronized (barCaches) {
				StatusBar.glBegin();
				for (int i = 0; i < size; i++) {
					StatusBar bar = barCaches.get(i);
					if (bar != null && bar.visible) {
						bar.createUI(g);
					}
				}
				StatusBar.glEnd();
			}
		}
	}

	@Override
	public void update(long elapsedTime) {
		if (!visible) {
			return;
		}
		int size = barCaches.size();
		if (size > 0) {
			synchronized (barCaches) {
				for (int i = 0; i < size; i++) {
					StatusBar bar = barCaches.get(i);
					if (bar != null && bar.visible) {
						bar.update(elapsedTime);
					}
				}
			}
		}
	}

	@Override
	public void setVisible(boolean v) {
		this.visible = v;
	}

	@Override
	public boolean isVisible() {
		return visible;
	}

	@Override
	public RectBox getCollisionBox() {
		return null;
	}

	@Override
	public LTexture getBitmap() {
		return null;
	}

	@Override
	public int getWidth() {
		return 0;
	}

	@Override
	public int getHeight() {
		return 0;
	}

	@Override
	public void dispose() {
		this.visible = false;
		int size = barCaches.size();
		for (int i = 0; i < size; i++) {
			StatusBar bar = barCaches.get(i);
			if (bar != null) {
				bar.dispose();
				bar = null;
			}
		}
		barCaches.clear();
		barCaches = null;
	}
}
